tools/doc: improvements for types

silverwind opened the pull request:
##### Checklist
- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] documentation is changed or added
- [x] commit message follows commit guidelines

##### Affected core subsystem(s)
The first commit capitalizes all primitive types so the parser can pick them up.
The second commit adds logic to the doctool to allow parsing of `Type[]` array syntax.

#### Before
<img width="857" src="https://cloud.githubusercontent.com/assets/115237/22620107/756927ea-eb04-11e6-9ae6-05a24ecf0d89.png">

#### After
<img width="874" src="https://cloud.githubusercontent.com/assets/115237/22620110/7e05c71e-eb04-11e6-9393-e4c3432b7108.png">

TimothyGu commented:
The first commit LGTM. For the second, I would prefer a more generic syntax like `Array<String>`. This way, we can not only have links to MDN for both Array and String, but also have better symmetry with other language types/protocols like Promise and Iterable (e.g. for `URLSearchParams` it'd be `Iterable<String, String>`).

mscdex commented:
I actually prefer primitives to stay lowercase. Also, in the particular example you showed, the inline 'pem' property description should actually have `Buffer` instead of `buffer` too.

silverwind commented:
> I actually prefer primitives to stay lowercase

I don't have stats, but I estimate that > 80% of our types are currently capitalized.

> the inline 'pem' property description should actually have Buffer instead of buffer too.

That'd be quite hard to parse. It's not following the type convention of `{Type}` at all.

mscdex commented:
> That'd be quite hard to parse. It's not following the type convention of {Type} at all.

I'm not suggesting it should parse it necessarily, I'm mostly just pointing out a typo that should be corrected.

lpinca commented on tools/doc/type-parser.js:
```diff
@@ -62,6 +62,17 @@ module.exports = {
           typeUrl = nodeDocUrl + typeMap[typeText];
         }
 
+        if (/\[]$/.test(typeText)) {
```
Can we merge this `if...else` chain with the previous one?

lpinca commented on tools/doc/type-parser.js:
```diff
@@ -62,6 +62,17 @@ module.exports = {
 
+        if (/\[]$/.test(typeText)) {
+          const typeTextBare = typeText.slice(0, -2);
+          if (jsPrimitives[typeTextBare]) {
```
Can we merge this `if...else` chain with the previous one on line 57?

silverwind commented:
So on current master we have 571 occurences in docs with capitalized primites and 160 without. I could change those all to lowercase as suggested by @mscdex. MDN currently lists them capitalized and the `href` would need to be capitalized as well.

thefourtheye commented:
> ... the href would need to be capitalized as well.

I would consider this as a problem.

silverwind commented:
Updated:

- Primitive types are now all lowercased. This seems to be a common convention in JS documention.
- Removed some duplicated code as per @lpinca's suggestion and made sure primitives of any case are picked up correctly.

@nodejs/documentation PTAL.

sam-github commented:
Consistency sounds great to me.

Is the example in https://github.com/nodejs/node/pull/11167#issuecomment-279829648 what it looks like now?

How come `string` is lower case, and `Buffer` and `Object` are upper case? It seems a bit random. Its not even built-in vs node specific, because String and Object are both builtin.

silverwind commented:
Because `string`is a [primitive](https://developer.mozilla.org/en-US/docs/Glossary/Primitive) while the others are not.

TimothyGu commented:
@silverwind, have you seen my comment above?
 https://github.com/nodejs/node/pull/11167#issuecomment-277466230

silverwind commented:
Yes, I think `Array<String>` make sense for non-Array types, and it's certainly something we can support in the parser. I'd just like to avoid packing too many changes in this PR. Maybe open an separate issue for it?

silverwind commented:
While I couldn't find definitive recommendations, the style of lowercasing primitives (and capitalizing everything else) is in use by both JSDoc and closure compile. See examples [here](http://usejsdoc.org/tags-type.html) and [here](https://github.com/google/closure-compiler/wiki/Annotating-JavaScript-for-the-Closure-Compiler).

sam-github commented:
Not so sure that's a useful distinction that we need to convey to our doc readers, given that strings do have methods in practical cases, but at least its something we can point to in our docs, and say we are following jsdoc.

Did you mean to write `Array<String> `, not `Array<string>`?

joyeecheung commented:
BTW [flow](https://flowtype.org/docs/quick-reference.html) and [TypeScript](https://www.typescriptlang.org/docs/handbook/basic-types.html) are using lower-cased primitive types too(therefore used by [definitely typed](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/node/index.d.ts), where a lot to people get their types from in their tools nowadays)

TimothyGu commented:
FWIW, [Tern](https://ternjs.net/) also uses lowercased primitive types.

thefourtheye commented:
My mind is trained to understand a name as a constructor function if it starts with a capital letter. So I am going to go with lowercase names.

silverwind commented:
I think we are in agreement that lowercase primitives is the way to go. Can I get a few LGTMs? I'll rebase and land shortly after it's approved.

silverwind commented:
Thanks! Landed in https://github.com/nodejs/node/commit/ff1361957df58ef92c8a9dfce40c6091ea3a21d4 and https://github.com/nodejs/node/commit/9be03a2903a0196212868d9e4282d638355828e7.

silverwind commented:
Whoops, PR incoming.

silverwind commented:
https://github.com/nodejs/node/pull/11658

evanlucas commented:
This is not landing cleanly on v7.x-staging. Mind submitting a backport PR?

silverwind commented:
Will do once https://github.com/nodejs/node/pull/11697 lands.

MylesBorins commented:
do we want to backport to v6.x?

MylesBorins commented:
ping

silverwind commented:
Backport in https://github.com/nodejs/node/pull/13054
